### ISSUE TITLE
minor issue about F# lexer

### ISSUE THREAD

**Issue by zufuliu** [ScintillaOrg/lexilla#147]
Hi @rdipardo I think it's better to move following `set*` and `numeric*` variables into `LexerFSharp` class (similar to other lexers) to reduce global constructor / destructor.
https://github.com/ScintillaOrg/lexilla/blob/a97769656d85de38ed978c13c0d2c30122348719/lexers/LexFSharp.cxx#L94-L101

drag lexilla.dll onto peview.exe (available at https://systeminformer.sourceforge.io/nightly.php), and see the CRT section:
![image](https://user-images.githubusercontent.com/2289926/223696426-eb284bd2-6880-48b9-ba38-e84b7a253332.png)

**Comment by rdipardo** [comment]
How do you demangle the symbols? All it shows me is `"GetNameSpace+0x<offset>"`.

<img width="408" alt="crt_init_bloat" src="https://user-images.githubusercontent.com/59004801/223805417-20462f31-2549-45bb-b54d-e10d36240b16.png">

**Comment by zufuliu** [comment]
> How do you demangle the symbols?

I just build locally with `nmake -f lexilla.mak` under "x64 Native Tools Command Prompt for VS 2022".

**Comment by rdipardo** [comment]
@zufuliu,
See if this patch produces a slimmer binary: [0001-Issue-147.patch.txt](https://github.com/ScintillaOrg/lexilla/files/10925808/0001-issue-147.patch.txt)

@nyamatongwe,
The overall change is +5 lines, so this may be a good time to shrink some lingering redundancies for a more maintainable net result: [0002-Prefer-NSDM-initialization-delete-copy-ctor.patch.txt](https://github.com/ScintillaOrg/lexilla/files/10925809/0002-Prefer-NSDM-initialization-delete-copy-ctor.patch.txt)


The *optional* second patch:

  + uses non-static data member initialization lists in place of constructors in some trivial structs
  + explicitly deletes the copy constructor and assignment operator overloads, mainly to appease MSVC, which (since about [version 16.11.1]) was already deleting them anyway (even at '/O1' optimization level):

        lexers\LexFSharp.cxx(360): warning C4625: '`anonymous-namespace'::LexerFSharp': copy constructor was implicitly defined as deleted
        lexers\LexFSharp.cxx(360): warning C4626: '`anonymous-namespace'::LexerFSharp': assignment operator was implicitly defined as deleted

  + resolves one of MSVC's C4365 warnings, previously hidden by a suppression for a [known issue] with [`CharacterSetArray::AddString`]




[version 16.11.1]: https://learn.microsoft.com/en-us/visualstudio/releases/2019/release-notes#16.11.1
[known issue]: https://groups.google.com/g/scintilla-interest/c/OHpjNv5YvzI/m/abgWUn6dBQAJ
[`CharacterSetArray::AddString`]: https://github.com/ScintillaOrg/lexilla/blob/a97769656d85de38ed978c13c0d2c30122348719/lexlib/CharacterSet.h#L48

**Comment by zufuliu** [comment]
The patch works:
![image](https://user-images.githubusercontent.com/2289926/223875862-59a0d7df-f9be-4d5b-91ff-e6b4e23d3b48.png)

**Comment by nyamatongwe** [comment]
For patch 2, to delete the 4 copy/move * constructor/assignment methods, the non-const arguments should be `&&` for move constructor and move assignment.

**Comment by zufuliu** [comment]
> All it shows me is `"GetNameSpace+0x<offset>"`

It turns out the associated pdb (lexilla.pdb for lexilla.dll) is required to resolve symbol name.
![image](https://user-images.githubusercontent.com/2289926/224004654-00609c14-1955-4e51-b4aa-3810f8927314.png)

**Comment by rdipardo** [comment]
> It turns out the associated pdb . . . is required

Got it.

> For patch 2, . . . the non-const arguments should be `&&` for move constructor and move assignment.

I haven't the skills (nor, it appears, the tooling) to assess the semantic correctness, but here's an amended version:

[0003-Declare-deleted-move-ctor-with-Lvalue.txt](https://github.com/ScintillaOrg/lexilla/files/10935675/0003-Declare-deleted-move-ctor-with-Lvalue.txt)

**Comment by nyamatongwe** [comment]
> I haven't the skills (nor, it appears, the tooling) to assess the semantic correctness

Visual C++ Code Analysis has:
~~~
C:\u\hg\lexilla\lexers\LexFSharp.cxx(292): warning C26432:
If you define or delete any default operation in the type 'class `anonymous-namespace'::LexerFSharp', define or delete them all (c.21).
~~~

Clang-Tidy (which can be easily used from Qt Creator) has
~~~
C:\u\hg\lexilla\lexers\LexFSharp.cxx:263:7: warning:
class 'LexerFSharp' defines a non-default destructor, a copy constructor and a copy assignment operator but does not define a move constructor or a move assignment operator [cppcoreguidelines-special-member-functions]
~~~

These both disappear after [0003-Declare-deleted-move-ctor-with-Lvalue.txt](https://github.com/ScintillaOrg/lexilla/files/10935675/0003-Declare-deleted-move-ctor-with-Lvalue.txt).

I don't normally mention these to implementers as they are part of the intimidating complexity that turns people away from C++.

**Comment by nyamatongwe** [comment]
Committed changes.

When possible, its helpful to include the issue number at the start of the commit message like `#147 ...` as that links the issue and commits for navigation. I try to remember to do this when committing but sometimes forget.

**Comment by nyamatongwe** [comment]
Included in version 5.2.4.
